xilem_core/src/views/adapt.rs

```diff
@@ -0,0 +1,186 @@
+use core::marker::PhantomData;
+
+use crate::{DynMessage, MessageResult, Mut, View, ViewId, ViewPathTracker};
+
+/// A view that wraps a child view and modifies the state that callbacks have access to.
+///
+/// # Examples
+///
+/// Suppose you have an outer type that looks like
+///
+/// ```ignore
+/// struct State {
+///     todos: Vec<Todo>
+/// }
+/// ```
+///
+/// and an inner type/view that looks like
+///
+/// ```ignore
+/// struct Todo {
+///     label: String
+/// }
+///
+/// struct TodoView {
+///     label: String
+/// }
+///
+/// enum TodoAction {
+///     Delete
+/// }
+///
+/// impl View<Todo, TodoAction, ViewCtx> for TodoView {
+///     // ...
+/// }
+/// ```
+///
+/// then your top-level action (`()`) and state type (`State`) don't match `TodoView`'s.
+/// You can use the `Adapt` view to mediate between them:
+///
+/// ```ignore
+/// state
+///     .todos
+///     .enumerate()
+///     .map(|(idx, todo)| {
+///         Adapt::new(
+///             move |data: &mut AppState, thunk| {
+///                 if let MessageResult::Action(action) = thunk.call(&mut data.todos[idx]) {
+///                     match action {
+///                         TodoAction::Delete => data.todos.remove(idx),
+///                     }
+///                 }
+///                 MessageResult::Nop
```

Comment:
I don't *think* `Nop` is right here if the `MessageResult` was `Action`. But, as I say in my comments on those types, the semantics have very much not been nailed down.

I suspect we might need a `map_action` method on `MessageResult` which does this properly?

Reply:
Same as below, but I agree as well, a `map_action` is probably better and reduces boilerplate as well

Reply:
> I suspect we might need a map_action method on MessageResult which does this properly?

btw. that already existed as `MessageResult::map` should I just port that?

Reply:
> I don't think `Nop` is right here if the `MessageResult` was Action

On a second look, I do think it is? I think anything else should probably be redirected to the parent view, but when the `Adapt` has handled everything in `MessageResult::Action`, it's a `Nop` for the parent view.

But yes semantics need to be nailed down at some point. 
E.g. if we want something like `MessageResult::RequestRebuild` but I think, unless we want to constrain `Action` with a trait bound (in case it has to be mapped and shouldn't just be a `Nop`), we probably want some "hardcoded" variants.

Reply:
In my mind, what `Nop` signifies is that this message did nothing; therefore, the View tree does not need to be rebuilt, because `State` wasn't modified

Reply:
Ah, I see that bit of logic in xilem(_masonry) now, and can follow your thoughts a little bit more.
So previously in old (async) xilem, every message/event resulted in a full rebuild, only when an async `Wake` event was sent (that was given to the view via `Cx::waker()`), it listened to `MessageResult::RequestRebuild` to check whether the async view needs (another) rebuild.

I'm not sure how safe it is to not rebuild when `MessageResult` is `Nop`, I think we need to be sure that the user had no real access to the `AppState` in that case.

Reply:
It seems reasonable that a developer would want to filter down messages they are interested.

The best example for that is probably the `button_any_state` we currently have, where you would send a `Nop` reply for any button presses you aren't interested in (but yes, filtering by `PointerButtons` for this case is *better* anyway)

Reply:
Yeah good point, though I think it *can* result in "unexpected" bugs when the user gets more power over the rebuild process (which should not be too expensive anyway).

Probably a good point for discussion in office hours, I think Raph had some thoughts to that as well.